pkg/broker/ingress/handler.go

```diff
@@ -18,97 +18,152 @@ package ingress
+	Send(ctx context.Context, ns, broker string, event cev2.Event) protocol.Result
+}
+
+// HttpMessageReceiver is an interface to listen on http requests.
+type HttpMessageReceiver interface {
+	StartListen(ctx context.Context, handler nethttp.Handler) error
 }
 
 // handler receives events and persists them to storage (pubsub).
+// TODO(liu-cong) add metrics
+// TODO(liu-cong) add tracing
+// TODO(liu-cong) support event TTL
 type handler struct {
-	// inbound is the cloudevents client to receive events.
-	inbound cev2.Client
+	// httpReceiver is an HTTP server to receive events.
+	httpReceiver HttpMessageReceiver
 	// decouple is the client to send events to a decouple sink.
 	decouple DecoupleSink
+	logger   *zap.Logger
 }
 
-func NewHandler(ctx context.Context, options ...Option) (*handler, error) {
-	h := &handler{}
-	for _, option := range options {
-		option(h)
+// NewHandler creates a new ingress handler.
+func NewHandler(ctx context.Context, options ...HandlerOption) (*handler, error) {
+	h := &handler{
+		logger: logging.FromContext(ctx),
 	}
 
-	var err error
-	if h.inbound == nil {
-		if h.inbound, err = newDefaultHTTPClient(); err != nil {
-			return nil, fmt.Errorf("failed to create inbound cloudevent client: %w", err)
+	for _, option := range options {
+		if err := option(h); err != nil {
+			return nil, err
 		}
 	}
 
+	if h.httpReceiver == nil {
+		h.httpReceiver = kncloudevents.NewHttpMessageReceiver(defaultPort)
+	}
+
 	if h.decouple == nil {
-		if h.decouple, err = h.newDefaultPubSubClient(ctx); err != nil {
-			return nil, fmt.Errorf("failed to create decouple cloudevent client: %w", err)
+		sink, err := NewMultiTopicDecoupleSink(ctx)
+		if err != nil {
+			return nil, err
 		}
+		h.decouple = sink
 	}
 
 	return h, nil
 }
 
+// Start blocks to receive events over HTTP.
 func (h *handler) Start(ctx context.Context) error {
-	return h.inbound.StartReceiver(ctx, h.receive)
+	return h.httpReceiver.StartListen(ctx, h)
 }
 
-// receive receives events from inbound and sends them to decouple.
-func (h *handler) receive(ctx context.Context, event cev2.Event) protocol.Result {
-	// TODO(liu-cong) add metrics
-	// TODO(liu-cong) route events based on broker.
+// ServeHTTP implements net/http Handler interface method.
+// 1. Performs basic validation of the request.
+// 2. Parse request URL to get namespace and broker.
+// 3. Convert request to event.
+// 4. Send event to decouple sink.
+func (h *handler) ServeHTTP(response nethttp.ResponseWriter, request *nethttp.Request) {
+	h.logger.Debug("Serving http", zap.Any("request", request))
+	if request.Method != nethttp.MethodPost {
+		response.WriteHeader(nethttp.StatusMethodNotAllowed)
+		return
+	}
 
-	if res := h.decouple.Send(ctx, event); !cev2.IsACK(res) {
-		logging.FromContext(ctx).Error("Error publishing to PubSub", zap.String("event", event.String()), zap.Error(res))
-		return res
+	// Path should be in the form of "/<ns>/<broker>".
+	pieces := strings.Split(request.URL.Path, "/")
```

Comment:
I think it would be better to use something like gorilla/mux for routing, especially since we might also want to serve zpages on the same port.

Reply:
I'd like to try that. Is it OK address this as a follow up https://github.com/google/knative-gcp/issues/785?

Reply:
Sounds good